client/my-sites/stats/wordads-chart-tabs/index.jsx

```diff
@@ -0,0 +1,252 @@
+				break;
+			case 'year':
+				dateLabel = date.format( 'YYYY' );
+				break;
+		}
+
+		tooltipData.push( {
+			label: dateLabel,
+			className: 'is-date-label',
+			value: null,
+		} );
+
+		switch ( this.props.chartTab ) {
+			default:
+				tooltipData.push( {
+					label: this.props.translate( 'Ads Served' ),
+					value: this.props.numberFormat( item.data.impressions ),
+					className: 'is-impressions',
+					icon: 'visible',
+				} );
+				tooltipData.push( {
+					label: this.props.translate( 'Avg. CPM' ),
+					value: '$ ' + this.props.numberFormat( item.data.cpm, { decimals: 2 } ),
+					className: 'is-cpm',
+					icon: 'stats-alt',
+				} );
+				tooltipData.push( {
+					label: this.props.translate( 'Revenue' ),
+					value: '$ ' + this.props.numberFormat( item.data.revenue, { decimals: 2 } ),
+					className: 'is-revenue',
+					icon: 'money',
+				} );
+				break;
+		}
+
+		return tooltipData;
+	}
+
+	onLegendClick = chartItem => {
+		const activeLegendCharts = this.state.activeLegendCharts;
+		const chartIndex = activeLegendCharts.indexOf( chartItem );
+		let gaEventAction;
+		if ( -1 === chartIndex ) {
+			activeLegendCharts.push( chartItem );
+			gaEventAction = ' on';
+		} else {
+			activeLegendCharts.splice( chartIndex );
+			gaEventAction = ' off';
+		}
+		this.props.recordGoogleEvent(
+			'WordAds Stats',
+			`Toggled Nested Chart ${ chartItem } ${ gaEventAction }`
+		);
+		this.setState( {
+			activeLegendCharts,
+		} );
+	};
+
+	getActiveTab = memoize( chartTab => {
+		const activeTab =
+			find( this.props.charts, { attr: this.props.chartTab } ) || this.props.charts[ 0 ];
+		const activeCharts = activeTab.legendOptions ? activeTab.legendOptions.slice() : [];
+		if ( this.state && activeTab !== this.state.activeTab ) {
+			this.setState( {
+				activeLegendCharts: activeCharts,
+				activeTab: activeTab,
+			} );
+		}
+		return activeTab;
+	} );
+
+	getLoadedData() {
+		const { fullQueryData } = this.props;
+		return fullQueryData;
+	}
+
+	buildChartData() {
+		const data = this.getLoadedData();
+		if ( ! data ) {
+			return [];
+		}
+
+		const activeTab = this.props.chartTab;
+		const labelKey =
+			'label' +
+			this.props.period.period.charAt( 0 ).toUpperCase() +
+			this.props.period.period.slice( 1 );
+		return data.map( record => {
+			let recordClassName;
+			if ( record.classNames && record.classNames.length ) {
+				recordClassName = record.classNames.join( ' ' );
+			}
+
+			const className = classNames( recordClassName, {
+				'is-selected': record.period === this.props.queryDate,
+			} );
+
+			const item = {
+				label: record[ labelKey ],
+				value: record[ activeTab ],
+				data: record,
+				className,
+			};
+			item.tooltipData = this.buildTooltipData( item );
+
+			return item;
+		} );
+	}
+
+	render() {
+		const { query, fullQueryRequesting, siteId } = this.props;
+		const chartData = this.buildChartData();
+		const activeTab = this.getActiveTab( this.props.chartTab );
```

Comment:
When toggling between chart tabs on the new display, I'm getting the following console message:

```Warning: Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.```

Could that be related to my memoization here?

Reply:
That's because your `render` method calls `getActiveTab` it calls `this.setState` in turn. That's forbidden during render. If you look at the original code in `stats-chart-tabs`, the `getActiveTab` there just reads `props` and computes something from them. The `setState` happens only in `componentWillReceiveProps` when reacting to changes in props passed from above (namely, the active tab).

Note also that @jsnmoon completely refactored this logic in #27409 and moved the state up one component, to `StatsSite`. You'll need to update your code accordingly when rebasing your (rather old) branch onto the latest `master`.

Reply:
Also, memoizing `getActiveTab` shouldn't be needed at all. It's searching in array with 4 elements and the memoization overhead very likely makes that slower rather than faster 🙁

Reply:
that's okay, good opportunity for me to learn memoization :)